### DISCUSSION TITLE
Use relative URLs instead of absolute

### DISCUSSION THREAD

**Discussion by mikecardwell** [vercel/next.js#32216] [Ideas]
There are many tags in the HTML which NextJS generates which contain absolute URLs. For example:

```
href="/_next/blah"
```

I've added selenium + chrome headless testing to my Next project, and it works fine when I test http://localhost:3000, but when I do a `next export`, and then try to to test `file:///home/user/project_dir/out/index.html`, all of those absolute links are broken.I.e, it tries to fetch `file:///_next/blah` instead of: `file:///home/user/project_dir/out/_next/blah`.

If I post-edit the generated HTML to make those links relative, then it all works fine.

I need to be able to run my tests against the static files which are generated by export to ensure that they have been generated correctly before publishing.

I suspect this is also an issue when people simply wish to run their nextjs application from a location other than the root of their website.

- [x] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

## Your Environment

| Tech    | Version |
|---------|---------|
| next    | v3.0.1-beta.15 |
| node    | v8.1.3 |
| OS      | Debian 9 |
| browser | Google Chrome 59.0.3071.115 |

**Comment by slowdub** [comment]
@mikecardwell [Other](https://github.com/zeit/next.js/issues/2335) issues are talking about this. The solution is to use [assetPrefix](https://github.com/zeit/next.js#cdn-support-with-asset-prefix).

**Comment by mikecardwell** [comment]
That is not a very good solution as it would require me to build one set of files to be served by a webserver, and then a different set of files to be accessed via a file:/// url. A better solution would be the one I described where relative URLs are used throughout.

**Comment by slowdub** [comment]
@mikecardwell I meant that you can use `assetPrefix` to achieve your described solution (create relative paths):
`assetPrefix: './'`

But I still can't see why is it required to preview locally via file URI? Couldn't you use [serve](https://github.com/zeit/serve) or any other http server locally?

**Comment by timneutkens** [comment]
Making the urls relative won't fix the issue. Since then you'd have issues when there's nested pages, `pages/some-dir/some-route.js` (`some-dir/some-route` route) would get a could not find `some-dir/some-route/_next/*`. The right solution is to use for example [serve](https://github.com/zeit/serve) to serve the html files through a webserver. That would be a better test anyway right, since users will get the html served through a webserver 🤔

**Comment by mikecardwell** [comment]
@timneutkens - Your example doesn't work because @Blistok's solution was wrong and would not work. When I say to use relative URLs, I mean that each html page which is generated would know where it is relative to the content it is trying to link to. For example when generating "some-dir/some-route", it would know that it is 2 levels deep, and it would know that the _next directory is one level deep, so when generating that particular page of HTML, it would use "../_next/blah". When generating another page for the same site, it might need to use "../../../_next/blah". This isn't just the "_next" directory I'm talking about either, it's all resources, e.g. link tags etc.

As it stands, "next export" doesn't produce a website which can be served from anywhere, it spits out a website that can be served from a particular subset of URL structures.

Here is a real life example of a website where that is a problem: http://www.exim.org - That is a statically generated website. It also has a large number of mirrors - http://www.exim.org/mirmon/www_mirrors.html - Many of these mirror a website from a sub url. E.g http://mirror.easyname.at/exim/ - The mirrors are updated by simply rsyncing the static files from the source on a schedule. This setup could not work with the files that nextjs spits out, because they are not portable.

Here's another use case: I want to build a nextjs app which I can just export to disk, zip up and distribute it as an application for people to unzip on their desktops/laptops and open in their web browser using "File -> Open". I can't do that with next export as it stands, because everyone will be opening the project from a different directory.

HTML documentation that comes packaged along with the application in a .tgz... Step 1 for reading this documentation, install a web server?

If nextjs used relative urls then you wouldn't need to tell people who want to serve up their site on things like github pages to add assetPrefix to their nextjs configuration. It would "just work".

[edit] You closed this issue prematurely IMO. I hope that doesn't mean my response is ignored.

**Comment by olizilla** [comment]
Support for exporting a static site with internal links and asset references all being relative urls would be rad for statically hosting next.js apps on [ipfs](https://ipfs.io/) or, as @mikecardwell explains, any situation where the site needs to be self-contained and can't assume a specific mounting path.

**Comment by binary64** [comment]
+1 Will be like, "woah, man" once Mike's ideas become reality.

**Comment by TooTallNate** [comment]
@timneutkens I am re-opening this. It's causing Next.js apps that are routed through a Path Alias rule to return the incorrect hostname which leads to potential CSP issues, etc.

It would be nice to have a good solution for this.

**Comment by arunoda** [comment]
@TooTallNate we use path alias for `zeit/docs` without any issues using absolute URLs. I really don't like to use relative URLs since they complicate things unnecessary. 
The state is undefined when someone use a relative URL inside a component which used in multiple pages.

**Comment by mikecardwell** [comment]
@arunoda The router could be extended with a function to allow you to look up the current path to the root of the app. Then people could just wrap such components in the withRouter HOC and then do things like:

```javascript
<img src={ `${this.props.router.pathToRoot}/static/images/foo.png` }/>
```

Which would sometimes spit out `<img src="./static/images/foo.png"/>` and other times `<img src="../../../static/images/foo.png"/>` depending on the current URL.

Of course, people who are happy with absolute URLs can just continue doing what they're doing right now, but for all of the use cases which require relative URLs, a solution would exist.

**Comment by olizilla** [comment]
I'd be totally happy with a manual solution to this like @mikecardwell suggests.

**Comment by tanishq-dubey** [comment]
Gonna +1 this issue.

Perhaps this is a specific use case (but I have seen [others complaining](https://stackoverflow.com/questions/51411662/issue-with-kubernetes-ingress-routing-to-nextjs-applications)).

My situation is that I am running multiple nextJS apps in a Kubernetes cluster with a NGINX ingress module getting requests from an AWS ELB. Each app is assigned a namespace such as `mysite.com/app1` or `mysite.com/app2`. The NGINX ingress is configured such that when a request is received on `/app1`, it forwards it to the root (`/`) of the container running app1 (because `app1` is expecting requests on `/`). 

To make links work (see: `href`), the NGINX instance adds a [`base`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base) element to the `head` of the page, so that each URL (such as going to `/contact-us` on a example page) is prepended with our base, hence we get `/contact-us` => `mysite.com/app1/contact-us` and everyone is happy!

This of course, doesn't work with nextJS as the paths are all absolute instead of relative (`base` only works on relative paths) and since this `base` is being added by NGINX on requests, it is difficult to guess the URL during build (auto deploy URLs/Machine Generated URLS). 

Would love to see support for either `base`, relative paths, or perhaps a completely different solution!

**Comment by Yuripetusko** [comment]
Any solutions to this? I am running nextjs static export from cordova which doesn't have webserver and serves pages as `file://` so need to have relative paths

**Comment by mikecardwell** [comment]
I think the only current solution is to post process the exported HTML to modify the links.

**Comment by Yuripetusko** [comment]
Thanks, the main issue with pathanmes like `'/'` or `'/home'` imagine doing search and replace of string `'/'` not a pretty task

**Comment by mikecardwell** [comment]
Yeah, it's rubbish. I'd be tempted to leverage headless chrome via the Puppeteer library to modify the links.

I guess the alternative would be to modify NextJS to give it the support I described earlier in this thread. I don't think the core devs are interested in making those changes themselves.

**Comment by MatteoGioioso** [comment]
Same issue here, I have used `assetPrefix: './'` which fixed the `href` of the link tags. However my static assets, like images and videos links, are broken.

**Comment by revmischa** [comment]
An option to export as relative paths (just don't prefix everything with `/`) would be very desirable. I expect to just be able to open index.html of my exported files and see things load. What's the reason to make it absolute by default?

**Comment by Yuripetusko** [comment]
for cordova i solved this using cordova local webserver

**Comment by tky5622** [comment]
@Yuripetusko 
Great! I would be  happy if you could share sample project or code .
 At present, there is little information on next.js with cordova, 
 So it will be very very helpfull to anyone who want to use it.

**Comment by Yuripetusko** [comment]
@tky5622 ok good call, will write something more formal in next few days

**Comment by sebfie** [comment]
@Yuripetusko Can you explain fast how you did? :)

**Comment by ItalyPaleAle** [comment]
+1 on this issue

My use case is a bit different. I'd like to use Next to build an app and distribute that over IPFS. When an app is running through IPFS, it's impossible to know the path it will be running in, so using `assetPrefix` isn't an option.

For example, my app could be running in any of the following:

- `http://<gateway-address>/ipfs/<cid>/index.html` (where `<cid>` is the hash of the content)
- `http://<gateway-address>/ipns/<some-string>/index.html`
- `http://<some-domain>/index.html` when there's a domain serving the request (e.g. when using the Cloudflare IPFS gateway)

**Comment by nblthree** [comment]
Can't this be solved with the `base` tag?

**Comment by ItalyPaleAle** [comment]
No, because the base path is not always known or fixed. See my comment above.

**Comment by tanishq-dubey** [comment]
Also my comment from a couple months ago https://github.com/zeit/next.js/issues/2581#issuecomment-439286024 references the base tag.

**Comment by steve-dobbs** [comment]
I too am having this issue. I'm contracting for the government in Australia creating educational content using next, part of the requirement (antiquated though it may be) when we want to publish this educational website is that it should be a 'self-contained zip'. 

Is there no export option that can be added to make this 'just work'? Find and replace or custom post processing is cumbersome.

Also, it's strange that the documentation doesn't mention that a HTTP server is required. For next users like myself, it'd be great if you were more explicit about the fact that the output of next export is not self-contained and requires a server

**Comment by Yakima-Teng** [comment]
I too am having this issue. My case is that I need to place static files generated by next.js to server in Hybrid APP where html files are visited through file:// protocol and the absolute path is not a fixed value.

**Comment by Yuripetusko** [comment]
If you use ionic webview plugin then you can make it work since ionic runs local webserver serving your files to http://localhost:8081 then you just copy nextjs static export files to www folder and only export index.html and let other routes resolve using client side (SSR not needed)

**Comment by mmolinari** [comment]
Thank you @Yuripetusko, I can confirm it works for me. I just created a new Cordova project, added the plugin, copied the static files in `www/` and links work just fine. Something like this:
```
cordova create hello com.example.hello HelloWorld
cd hello/
cordova platform add ios
cordova plugin add https://github.com/ionic-team/cordova-plugin-ionic-webview

cd path/to/nextjs
npm run build && npm run export
cp -R out/* path/to/hello/www/

cd path/to/hello
cordova emulate
```
I also had to add `<preference name="ScrollEnabled" value="true" />` to `config.xml` because of https://github.com/ionic-team/cordova-plugin-ionic-webview#scrollenabled:
> Ionic apps work better if the WKWebView is not scrollable, so the scroll is disabled by default, but can be enabled with this preference. This only affects the main ScrollView of the WKWebView, so only affects the body, not other scrollable components.

I'm still testing, so it's possible there are other issues.

**Comment by Velenir** [comment]
To host on IPFS it is possible to use a hack where you dynamically inject `<base href="...">` tag at runtime:

```js
// ./pages/_document.js
import Document, { Html, Head, Main, NextScript } from 'next/document'

const scriptTxt = `
(function () {
  const { pathname } = window.location
  const ipfsMatch = /.*\\/Qm\\w{44}\\//.exec(pathname)
  const base = document.createElement('base')

  base.href = ipfsMatch ? ipfsMatch[0] : '/'
  document.head.append(base)
})();
`

class MyDocument extends Document {

  render() {
    return (
      <Html>
        <Head>
            <script dangerouslySetInnerHTML={{__html: scriptTxt}}/>
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}

export default MyDocument
```

There are also other issues that arise with it such as relative assetPrefix and routing, which all need their own workarounds.

You can see a small example of a Next.js app deployable to IPFS (with my current workarounds) here: https://github.com/Velenir/nextjs-ipfs-example/

**Comment by apnerve** [comment]
(slightly contrary opinion) I found the current behaviour more favourable for my flow where I keep on moving my components around while organising and using an absolute link means I can easily move the components that use `Link` without any issues.

**Comment by WoLfulus** [comment]
I don't know if this works for any case (I just have an index page), but this is what I ended up doing:

- Set `assetPrefix` to `assets` in config file
- `next build`
- `next export`
- `mv out/_next out/assets/_next`

Now I can serve the files anywhere I want to

🤷‍♀️

**Reply by MimusSD** [reply]
This is the same thing as just doing assetPrefix as ./

The problem is that this still breaks for hrefs in CSS files, which would now be looking in _next/static/css/assets/_next/...

**Comment by pawlarius** [comment]
still cant find any solutions after next js 9 in 2020 :((

**Comment by bhanupsingh77** [comment]
### How it worked for me you can also give it a try.

### Main points
1) Add next.config.js file with **module.exports = { assetPrefix: './' }** (you can have more properties in the module according to your code requirements )
2) Make change in package.json file in scripts with **"build": "next build && next export"**
3) Make changes to all next/link - Link herf value's (for eg. **index page -  earlier  (href="/") after change (href="./index.html")**  and **for all other pages like this earlier  (href="/PAGE_NAME") after change (href="./PAGE_NAME.html")**  

### --- imp
1) PAGE_NAME represent placeholder for your page name.
2) To access all your public folder content of next js for eg for image use src="./image.jpg" instead of "/image.jpg".
3) The "build": "next build && next export" in above Main point 2 will create a **out folder** with your build files that works ("build": "next build" that creates build files in .next folder won't work neither using next.config.js file with module.exports = { distDir: 'BUILD_FOLDER_NAME' } will not work) 
(BUILD_FOLDER_NAME is a placeholder)

**Comment by viztor** [comment]
I guess as of right now the issue is that the javascript code have no way of figuring out the base without extra configuration, after all the url are written as /xxx/yyy/zz instead of xxx#yyy/zzz, and because of that, it is difficult to figure out where the actual base path is just from the url itself.

**Comment by CavalloneChen** [comment]
In my case, this config helped:
```
module.exports = withPlugins([[withSass], [withImages]], {
  webpack(config, options) {
    config.resolve.modules.push(path.resolve("./"));
    return config;
  },
  basePath: "/xxx",
  assetPrefix: "/xxx"
});
```

**Comment by Codelica** [comment]
We find ourselves in a similar situation to @tanishq-dubey, where some path flexibility currently keeps us from being able to migrate to Next.

Currently we package our React apps as Docker images which allow for the `basePath` to be set at container startup (via env var). This is possible with some pretty minimal base path trickery via our current stack (create-react-app, React Router, Express), but doesn't seem to be possible with Next short of either:  A) leaving source in the image and making containers do a full re-build internally on startup if the basePath has changed or b) setting `basePath` to some unique token that a recursive sed run modifies as necessary before startup. If there is some cleaner option I'm overlooking I'd love to know, as both of those feel a little too dirty for us to consider. :) 

If we could get past the base path hangup, our only other issue is more of a nitpick.  Just that including the entire Next package in the image for serving the app feels pretty heavy at ~80MB uncompressed.  Would be nice if the production server could be broken out into its own module, but maybe that's crazy talk.  TBH I haven't researched the options much.

**Comment by j11s22k33** [comment]
18~ basepath

**Comment by erikian** [comment]
This feature would be great in Electron apps. In order to use Next.js with Electron in production, one would need to start a custom server on localhost, making it available in the browser as well, which is not always desirable. If this is implemented, we could just `appWindow.webContents.loadURL('/path/to/nextExportDir')`, completely eliminating the server overhead.

**Comment by bytebae** [comment]
Currently facing this in my first react project as well. Are we on the same boat or is there a method / workaround for this in react?

**Comment by mikehardenize** [comment]
I don't think there are any generally useable solutions/workarounds. I assume that, given the amount of demand for this functionality, and the lack of any movement on it in the 4 years this issue has been open, it's just too difficult for the devs to bolt on now. It should have been there from the beginning.

**Comment by yordis** [comment]
Kind of annoying situation honestly, `react-router@v6` is lighting years ahead of `next/router` except that it is missing the FS convention

**Comment by hidesminimally** [comment]
I'm running into this problem as well as I'm trying to figure out how to render my exported Next.js app within Electron.
I've come across a few options:
1. https://github.com/vercel/serve
2. https://github.com/sindresorhus/electron-serve 
3. https://github.com/HaNdTriX/next-electron-server

Has anyone evaluated whether these options are able to solve this problem?

**Comment by davidmurdoch** [comment]
Commenting just to say that this is such an obvious _default_ next.js should come with that I'm absolutely flabbergasted. Flabbergasted!

**Comment by msabramo** [comment]
This is really unfortunate. The requirement to use absolute URLs for assets prevents the Build Once, Run Anywhere promise of Docker from being realized. If you build a Docker image using built/exported Next.JS code, you end up with an image that only works when the app is hosted at a specific path. It is easily broken once you put it in an environment where it sits at a different place in the URL hierarchy, which is very common thing to do because people often want to have a reverse proxy that routes to different apps based on their path prefix.

**Reply by raeperd** [reply]
I am of the exact same opinion. In this respect, it is difficult to say that nextjs fully supports Docker.

**Comment by Leonardll** [comment]
i am facing the same wall as well. I have deployed my next app on fleek, everything is fine on the home page but others pages are breaking after refresh with an error message that says the path is incorrect.

**Reply by PatrickAlphaC** [reply]
Having similar issues with other decentralized hosting solutions like arweave. My IPFS stuff is working great though (cuz my IPFS stuff is really small).

**Reply by PatrickAlphaC** [reply]
Ended up getting it to work... by manually going through my exports and replacing them with relative imports.

**Reply by v1rtl** [reply]
@PatrickAlphaC it didn't work for me. Some styles and fonts fail to load by replacing to relative URL

**Comment by rwxguo** [comment]
I have the same requirement for using `file://` protocol to browser html pages offline, without any local web server.
Some people need this for testing, or for sharing the website to others through office network share drive (web server is not allowed) etc.
Most of the JamStack tools requires local web server for their experted static HTML, expect `MkDocs` I found so far.

My workaround is:

Instead of config `assetPrefix` in `next.config.js`, or using `<a>` instead of `next/link`, there is another workaround that using JavaScript to dynamically change each of the path in your HTML, when it detect is under `file://` protocol.

Here is the example, when page is laded, the script will check the protocol, and iterate all of `css` asset links, duplicate another one with their relative path:

```js
<script>
        if (window.location.protocol === 'file:') {
            const styleSheets = document.styleSheets;
            let head = document.getElementsByTagName('head')[0];
            for (let i = 0; i < styleSheets.length; i++) {
                let link = document.createElement('link');
                link.rel = 'stylesheet';
                link.type = 'text/css';
                link.href = `./${styleSheets[i].href.split('/').pop()}`;
                head.appendChild(link);
            }
        }
</script>
```

**Please notice, the script will cause infinite loop if you append the link with a workable href that is found and already loaded**

The same approach could also be applied to other assets or `<a>` href etc.

**Comment by csianglim** [comment]
I'm working on a static Next.js site and ran into this issue today. Being new to Next and coming from a Jekyll background, I'm surprised this is still a problem after 5+ years, Jekyll handles this out of the box with minimal configuration tweaks or awkward workarounds.

**Comment by DmacMcgreg** [comment]
Just had to urgently re-write an entire NextJS app project in Create React App with React Router because of this issue. file:// based routing is definitely still a thing with some large clients that distribute their app offline using secure intranet systems. Huge blow to the project, since we wanted to use a CMS with it using SSG, and this won't be possible with CRA. 

Would be a great addition for NextJS to have some sort of config to handle this type of use case.

Also, this would work for simple electron builds, which we build a lot of for conferences where we have to pass off a single executable to the conference team to run and display on a television. 

My opinion is that if Vercel is going to present NextJS as the best and universal way to build a react app, then it should support these valid use cases as well, not only the ones that directly benefit their business model. Ultimately, we are still your customers using vercel for various apps, but we have to build _these_ types of apps as well. And if that can't be done, then we have to look for alternatives and consider just making _all_ of our apps using those alternatives.

**Reply by benrudolphWSP** [reply]
Yes, especially government entities. Still very much a thing with them.

How did you go about converting to a React app? Was it easy? I'm in a bind and a client is beating down my door wondering why I can't make this work since the previous site did. I thought I was pushing this project forward moving to Next.js but simply not the case in this scenario.

**Comment by herbig** [comment]


**Reply by olizilla** [reply]
it'd be great if this were fixed in nextJS. In the meantime, the recommended way to fetch websites over IPFS is now to use a subdomain gateway like dweb.link where your site would be available at `https://<cid>.ipfs.dweb.link`. Putting the cid in the domain allows you to use `/` as the base URL for your site, and you avoid sharing an origin with every other website on that gateway. see: https://docs.ipfs.tech/how-to/address-ipfs-on-web/#subdomain-gateway

**Reply by olizilla** [reply]
for simpler static sites that don't rely on js for navigation, like those that `hugo` or `eleventy` would generate, you can run https://github.com/olizilla/all-relative on your generated output to convert all links in the html and css to their relative equivalent.

**Reply by herbig** [reply]
thank you @olizilla this is super helpful, I'll look into both these

**Comment by benrudolphWSP** [comment]
@mikecardwell Did you ever find a solution? I'm in the same boat where the site is being distributed via zip file and ran locally by opening in Windows Explorer. Total nightmare scenario as I previously had it working on a plain vanilla HTML/CSS/JS site and client is wondering why I can't make it work. Is it easy to convert project to a react app?

**Reply by mikecardwell** [reply]
I provided a simple solution 5 years ago in this comment - https://github.com/vercel/next.js/discussions/32216#discussioncomment-1766247 but that one requires some small changes to NextJS its self, which nobody has ever taken up.

So, I've just chosen not to use NextJS for projects where it doesn't work.

**Comment by DmacMcgreg** [comment]
You'll might just have to convert all the pages to React Router and CRA or
Vite if you want a quick solution. It might take a few hours depending on
the size of the app.

**Comment by dagoss** [comment]
+1 from me. I'm also running into this issue where I need to provide the client with a zip file that contains the application as standalone html/css/js and I can't control where or how it is deployed.

**Comment by mariotacke** [comment]
Just ran into this extremely annoying problem as well. Are there any plugins that could do this sort of transformation automatically? If this is possible through plugins, could you refer me to the relevant code sections?

**Comment by MimusSD** [comment]
Oh man is there no workaround for this? I can't even post-process the HTMLs because I have links in some CSS stylesheets to some media

**Comment by Otis-xp** [comment]
I wanted to package nextjs as an electron application, but nextjs couldn't export using relative paths, so the css and js files couldn't be found on the page of the packaged electron application

**Comment by benhansen-io** [comment]
I used logic similar to the following to get around this issue in an electron app:

```
  protocol.handle('file', (request) => {
    const url = request.url;

    const startOfNext = url.indexOf("/_next/");
    if (startOfNext == -1) {
      // Do not modify the request
      return net.fetch(url, {bypassCustomProtocolHandlers: true});
    }
    // Until nextjs can output static files that can be loaded from a directory (see https://github.com/vercel/next.js/discussions/32216)
    // we can either copy our files into the root directory or do this interception here to add the out directory.
    const new_url = [url.slice(0, "file://".length), app.getAppPath(), '/', REPORT_DIR, url.slice(startOfNext)].join('');
    return net.fetch(new_url, {bypassCustomProtocolHandlers: true});
  });
```

**Reply by gaurav21r** [reply]
@benhansen-io Great work but could you provide some context? Did you put this in the main process? Is `protocol` an export from `electron` ?

**Reply by benhansen-io** [reply]
That code is in the index.js of my electron app inside a

```
app.whenReady().then(() => {
...

})
```

and yes protocol is an electron export.

```
const { app, BrowserWindow, dialog, shell, protocol, net, contextBridge, ipcMain } = require('electron');
```

**Comment by ftm1337** [comment]
> hi, when we deploy a static site over github pages, it works with normal relative urls
but the same repository doesnt work on vercel.
Our goal is to enable "Downloadable Websites" for OFFLINE USAGE. It would be helpful if vercel can mention that its incapable of handling "simple" websites properly and caters only to rich websites that require paying for servers (either financially, or expending financial resources to teach non-tech people (>99% of current humans) about "serving websites with a webserver".

got told this just now by a volunteer building disaster relief content. i dont mean to be dramatic, but i hope you see their point. a simple solution would be simply enable a Framework Preset for static simple pages and Not "build" them. would also save some resources for vercel. or maybe just point that vercel is not able to support such pages and refer people to a competitor like github pages ;)

**Comment by bymagnum** [comment]
Export NextJS to the folder where Electron looks at when building a project or during development, you will get static files. Next, add your own scheme, for example, 'app', register it. Run your app fully. That's the whole solution. 


```
const scheme = 'app';

protocol.registerSchemesAsPrivileged([
        { scheme, privileges: { standard: true, secure: true, supportFetchAPI: true  } }
]);

const mainWindow = new BrowserWindow(browserWindowOpts);

mainWindow.loadURL(scheme  + '://localhost/index.html', loadOpts);
```
